sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java

```diff
@@ -158,48 +155,94 @@ public Mono<StoreResponse> invokeAsync(RxDocumentServiceRequest request,
 
     private Mono<StoreResponse> getStoreResponseMonoWithSpeculation(RxDocumentServiceRequest request, Quadruple<Boolean, Boolean, Duration, Integer> forceRefreshAndTimeout) {
         CosmosEndToEndOperationLatencyPolicyConfig config = request.requestContext.getEndToEndOperationLatencyPolicyConfig();
+        List<String> preferredRegions = this.transportClient.getGlobalEndpointManager().getPreferredRegions();
+        AvailabilityStrategy strategy = config.getAvailabilityStrategy();
+        if (strategy == null){
+            strategy = new DefaultAvailabilityStrategy();
+        }
+        if (strategy instanceof ThresholdBasedAvailabilityStrategy) {
+                speculativeProcessor = new ThresholdBasedSpeculation();
+        }
+
         List<Mono<StoreResponse>> monoList = new ArrayList<>();
         List<RxDocumentServiceRequest> requests = new ArrayList<>();
 
-        if (speculativeProcessor.shouldIncludeOriginalRequestRegion()) {
+        if (speculativeProcessor != null) {
+            List<String> effectiveRetryRegions = strategy.getEffectiveRetryRegions(preferredRegions,
+                request.requestContext.getExcludeRegions());
+            if (effectiveRetryRegions.size() > BridgeInternal.getNumberOfRegionsToTry(strategy)){
+                effectiveRetryRegions = effectiveRetryRegions.subList(0, BridgeInternal.getNumberOfRegionsToTry(strategy));
+            }
+            effectiveRetryRegions
```

Comment:
what would happen if no preferredRegions is being defined?

And also instead of preferredRegions, should we get writeRegion/readRegion instead?

Reply:
We get those while actually materializing the regions to send the requests